regtests/0336_random_task_safe/test.adb

```diff
@@ -0,0 +1,119 @@
+with Ada.Containers.Hashed_Sets;
+
+with Ada.Exceptions; use Ada.Exceptions;
+with Ada.Text_IO;    use Ada.Text_IO;
+with AWS.Utils;      use AWS.Utils;
+
+procedure Test is
+
+   type Test_Random_Mode is (None, Save, Lookup);
+
+   package Random_Sets is new Ada.Containers.Hashed_Sets
+     (Random_Integer, Ada.Containers.Hash_Type'Mod, "=");
+
+   Set : Random_Sets.Set;
+   Locker : RW_Semaphore (1);
+   Wrong_Data_Set : exception;
+
+   procedure Test_Random_Sequence
+     (Count : Positive; Mode : Test_Random_Mode);
+
+   function Test_Random_Sequence
+     (Initialor : Integer; Count : Positive) return String;
+
+   task Secondary is
+      entry Start (Count : Positive; Mode : Test_Random_Mode);
+      entry Done (Quit : Boolean);
+   end Secondary;
+
+   --------------------------
+   -- Test_Random_Sequence --
+   --------------------------
+
+   function Test_Random_Sequence
+     (Initialor : Integer; Count : Positive) return String is
+   begin
+      Random_Reset (Initialor);
+      Test_Random_Sequence (Count, None);
+      return Random_String (77);
+   end Test_Random_Sequence;
+
+   procedure Test_Random_Sequence
+     (Count : Positive; Mode : Test_Random_Mode)
+   is
+      R : Random_Integer;
+   begin
+      for J in 1 .. Count loop
+         R := Random;
+
+         case Mode is
+            when None =>
+               null;
+
+            when Save =>
+               Locker.Write;
+               Set.Include (R);
+               Locker.Release_Write;
+
+            when Lookup =>
+               if not Set.Contains (R) then
```

Comment:
shouldn't we call `Locker.write` before we read from the set ?

Reply:
Nope, we are not reading when writing. First phase is writing concurrently, second phase reading concurrently. Concurrent reading does not require protection.

Reply:
A few years ago, reading from containers was actually also changing some internal flags (hence the `pragma Suppress (Container_Checks)` that was introduced at the time. I forgot whether it has been fixed since then.

Reply:
When Lock and Busy counter became atomic incremented/decremented the cursor tempering check became RO task safe.

Reply:
thanks for confirming ! useful to know

Reply:
The only case when the RO operations with tempering check is not task safe, when platform does not support atomic increments. It this case we should suppress tempering check for RO concurrency available.